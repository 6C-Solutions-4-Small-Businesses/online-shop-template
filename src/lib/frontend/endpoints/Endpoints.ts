import type {CheckoutRequest, CheckoutResponse} from '$lib/frontend/requests/CheckoutRequest'
import {toastError} from '$lib/frontend/core/ToasterUtils'
import {Email, type FindUserRequest} from '$lib/frontend/requests/FindUserRequest'
import type {UserAccountSummaryPresentation} from '$lib/frontend/presentations/UserAccountSummaryPresentation'
import type {
    CustomerProfileEditionRequest,
    CustomerProfilePresentation
} from '$lib/frontend/presentations/CustomerProfilePresentation'
import type {OrderDetailsPresentation} from '$lib/frontend/presentations/OrderDetailsPresentation'
import {API_BASE_ENDPOINT, BASE_HEADERS} from '$lib/frontend/Constants'
import type {BusinessContactRequest} from '$lib/frontend/requests/BusinessContactRequest'
import type {GenericSuccessResponsePresentation} from '$lib/frontend/presentations/GenericSuccessResponsePresentation'

export async function createCheckoutSession(parameters: CheckoutRequest): Promise<CheckoutResponse | undefined> {
    const response = await fetch(
        `${API_BASE_ENDPOINT}/cart/checkout`,
        {
            method: 'POST',
            headers: BASE_HEADERS,
            body: JSON.stringify(parameters),
        }
    )

    if (response.status === 200) {
        return await response.json()
    } else {
        const message = await response.text()
        console.log(message)
        toastError(message)
    }
}

export async function findUser(parameters: FindUserRequest): Promise<UserAccountSummaryPresentation | undefined> {

    const response = await fetch(`${API_BASE_ENDPOINT}/user/find`, {
        method: 'POST',
        headers: BASE_HEADERS,
        body: JSON.stringify(
            parameters instanceof Email
                ? {email: parameters.value}
                : {phoneNumber: parameters.value}
        ),
    })

    switch (response.status) {
        case 200:
            return await response.json()
        case 404:
            return undefined
        case 500:
            toastError(await response.text())
    }
}

export async function finalizeCheckoutSession(sessionId: string): Promise<OrderDetailsPresentation | undefined> {
    const response = await fetch(
        `${API_BASE_ENDPOINT}/cart/checkout/${sessionId}`,
        {
            method: 'POST',
            headers: BASE_HEADERS,
        }
    )

    if (response.status === 200) {
        return await response.json()
    } else {
        toastError(await response.text())
    }
}

export async function createCustomerProfile(
    request: CustomerProfileEditionRequest
): Promise<CustomerProfilePresentation | undefined> {
    const response = await fetch(`${API_BASE_ENDPOINT}/user/customerProfile`, {
        method: 'POST',
        headers: BASE_HEADERS,
        body: JSON.stringify(request),
    })

    switch (response.status) {
        case 200:
        case 201:
            return await response.json()
        case 500:
        default:
            toastError(await response.text())
    }
}

export async function contactBusiness(
    request: BusinessContactRequest
): Promise<GenericSuccessResponsePresentation | undefined> {
    const response = await fetch(`${API_BASE_ENDPOINT}/contact/owner`, {
        method: 'POST',
        headers: BASE_HEADERS,
        body: JSON.stringify(request),
    })

    switch (response.status) {
        case 200:
        case 201:
            return await response.json()
        case 500:
        default:
            toastError('Malheureusement, nous n\'avons pas pu envoyer votre message. Veuillez r√©essayer plus tard.')
    }
}
