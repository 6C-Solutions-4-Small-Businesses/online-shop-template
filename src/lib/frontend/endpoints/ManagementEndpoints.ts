import {API_BASE_ENDPOINT, BASE_HEADERS} from '$lib/frontend/Constants'
import type {OffersManagementRequest} from '$lib/frontend/requests/OffersManagementRequest'
import type {OffersManagementResultPresentation} from '$lib/frontend/presentations/OffersManagementResultPresentation'

export async function sendOffersManagementRequests(request: OffersManagementRequest): Promise<OffersManagementResultPresentation | undefined> {
    const response = await fetch(`${API_BASE_ENDPOINT}/management/offers`, {
        method: 'POST',
        headers: BASE_HEADERS,
        body: JSON.stringify(request),
    })

    if (response.status === 201) {
        return await response.json()
    }
}
