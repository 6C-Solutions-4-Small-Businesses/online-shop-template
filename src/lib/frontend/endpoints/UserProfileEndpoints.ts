import type {UserProfilePresentation} from '$lib/frontend/presentations/UserProfilePresentation'
import type {UserAccountSummaryPresentation} from '$lib/frontend/presentations/UserAccountSummaryPresentation'
import type {AddressPresentation} from '$lib/frontend/presentations/AddressPresentation'

export async function getUserProfile(): Promise<UserProfilePresentation | undefined> {
    //TODO(https://github.com/6C-Solutions-4-Small-Businesses/static-web-site/issues/274)
    return Promise.resolve({
        account: {
            id: 'usr_1234567890',
            firstName: 'John',
            lastName: 'Doe',
            email: 'account@mail.com',
            phoneNumber: '5148907685',
            customerId: 'cus_XyX',
        } as UserAccountSummaryPresentation,
        deliveryAddress: {
            streetAddressLine1: '1234 Main St',
            streetAddressLine2: '',
            city: 'Anytown',
            state: 'CA',
            postalCode: '12345',
            country:'USA',
        } as AddressPresentation,
        billingAddress: {
            streetAddressLine1: '1234 Main St',
            streetAddressLine2: '',
            city: 'Anytown',
            state: 'CA',
            postalCode: '12345',
            country:'USA',
        } as AddressPresentation,
        orders: {
            completed: [],
            current: [],
            canceled: [],
        },
    } as UserProfilePresentation)
}
