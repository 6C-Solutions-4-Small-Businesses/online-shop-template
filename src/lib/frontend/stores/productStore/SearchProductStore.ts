import type {OfferSummaryPresentation} from '$lib/frontend/presentations/OfferSummaryPresentation'
import {writable, type Writable} from 'svelte/store'
import type {PaginationPresentation} from '$lib/frontend/core/PaginationPresentation'

export const searchedProductName: Writable<string> = writable()

export const submittedProductName: Writable<string> = writable()

export const searchedProductResult: Writable<PaginationPresentation<OfferSummaryPresentation>> =
    writable()
