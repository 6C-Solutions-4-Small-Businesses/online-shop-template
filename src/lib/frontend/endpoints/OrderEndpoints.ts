import type {OrderDetailsPresentation} from '$lib/frontend/presentations/OrderDetailsPresentation'
import {API_BASE_ENDPOINT, BASE_HEADERS} from '$lib/frontend/Constants'
import {EndpointError} from '$lib/frontend/endpoints/EndpointError'

export async function getOrderDetails(id: string): Promise<OrderDetailsPresentation> {
    const response = await fetch(
        `${API_BASE_ENDPOINT}/order/${id}`,
        {
            method: 'GET',
            headers: BASE_HEADERS
        }
    )

    switch (response.status) {
        case 200:
            return await response.json()
        default:
            throw new EndpointError(await response.text())
    }
}

