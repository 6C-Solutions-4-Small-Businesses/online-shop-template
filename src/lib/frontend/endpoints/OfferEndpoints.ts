import {API_BASE_ENDPOINT, BASE_HEADERS} from '$lib/frontend/Constants'
import {searchedProductResult} from '$lib/frontend/stores/productStore/SearchProductStore'

export async function fetchSearchedOffersResult(
    productName: string,
    page?: number
): Promise<void> {
    const response = await fetch(
        `${API_BASE_ENDPOINT}/offer?productName=${productName}&${page ? `page=${page}&` : ''}limit=18`,
        {
            method: 'GET',
            headers: BASE_HEADERS
        }
    )

    searchedProductResult.set(await response.json())
}
