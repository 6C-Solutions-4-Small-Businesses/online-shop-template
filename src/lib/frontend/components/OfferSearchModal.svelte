<script lang="ts">

    import {getModalStore} from '@skeletonlabs/skeleton'
    import BaseModal from '$lib/frontend/components/BaseModal.svelte'
    import {type SvelteComponent} from 'svelte'
    import SearchInput from '$lib/frontend/components/SearchInput.svelte'
    import {fetchSearchedOffersResult} from '$lib/frontend/endpoints/OfferEndpoints'
    import {searchedProductResult} from '$lib/frontend/stores/productStore/SearchProductStore'
    import type {OfferSummaryPresentation} from '$lib/frontend/presentations/OfferSummaryPresentation'
    import OfferSummaryCard from '$lib/frontend/components/OfferSummaryCard.svelte'
    import {throttle} from '$lib/frontend/core/RxHelper'

    const modalStore = getModalStore()
    const throttleSearchFunction = throttle(async (searchTerm: string) => {
        const searchedOffersResult = await fetchSearchedOffersResult(searchTerm)
        if (searchedOffersResult) {
            matchedOffers = searchedOffersResult.items
        } else {
            matchedOffers = []
        }
    }, 500)

    export let parent: SvelteComponent

    let modalSettings = $modalStore[0]
    let matchedOffers: OfferSummaryPresentation[] = $searchedProductResult?.items
    let searchTerm: string = ''

    async function handleSearchTermChange(value: string): Promise<void> {
        searchTerm = value
        const func = await throttleSearchFunction
        func(searchTerm)
    }

    async function onSearchResetHandler() {
        matchedOffers = []
        searchTerm = ''
    }

</script>
{#if modalSettings}
    <div class="w-full md:w-6/12 lg:w-6/12 xl:w-4/12">
        <BaseModal
                title={modalSettings.title}
                body={modalSettings.body}
                parent={parent}
                background="bg-gray-100"
                submitButtonDisabled={false}
                onSubmitClickHandler={() => Promise.resolve()}
                onCancelClickHandler={modalSettings.meta?.onClose}
        >
            <div class="w-full px-4">
                <SearchInput
                        inputId="search-input"
                        buttonId="search-submit-button"
                        width="w-full"
                        onSearchTermChangeHandler="{handleSearchTermChange}"
                        onSearchResetHandler="{onSearchResetHandler}"
                />
                <div class="py-4">
                    {#if searchTerm}
                        {#if matchedOffers?.length > 0}
                            <div class="w-full h-96 flex flex-col gap-2 overflow-y-scroll">
                                {#each matchedOffers.slice(0, 5) as offer}
                                    <OfferSummaryCard
                                            id="{offer.id}"
                                            name="{offer.name}"
                                            image="{offer.image}"
                                            price="{offer.price}"
                                            onClickHandler="{() => modalSettings?.response?.call({}, offer)}"
                                    />
                                {/each}
                            </div>
                        {:else}
                            <p class="text-center text-gray-700" data-testid="no-match-message"><span>{@html `Aucune offre trouvée pour "<b>${searchTerm}</b>"`}</span></p>
                        {/if}
                    {:else}
                        <p class="text-center text-gray-700" data-testid="description-message"><span>Cherchez l'offre que vous désirez modifier.</span></p>
                    {/if}
                </div>
            </div>
        </BaseModal>
    </div>
{/if}
