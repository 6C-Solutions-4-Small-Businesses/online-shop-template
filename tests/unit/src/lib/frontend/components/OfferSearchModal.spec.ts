import {afterEach, assert, beforeEach, describe, expect} from 'vitest'
import '@testing-library/jest-dom'
import {fireEvent, render, type RenderResult, screen as vScreen, waitFor} from '@testing-library/svelte'
import type {SvelteComponent} from 'svelte'
import {mock} from 'vitest-mock-extended'
import type {ModalStore} from '@skeletonlabs/skeleton'
import {currentModalSettings, currentModalStoreOnResponse} from '$mocks/src/lib/frontend/stores/ModalStore'
import OfferSearchModal from '$lib/frontend/components/OfferSearchModal.svelte'
import {searchedProductResult} from '$lib/frontend/stores/productStore/SearchProductStore'
import type {OfferSummaryPresentation} from '$lib/frontend/presentations/OfferSummaryPresentation'
import type {PaginationPresentation} from '$lib/frontend/core/PaginationPresentation'
import { fetchSearchedOffersResult } from '$mocks/src/lib/frontend/endpoints/OfferEndpoints'

describe('OfferSearchModal component', () => {

    const title = 'Title'
    const body = `Body`
    const onClose = vi.fn()

    let searchInput: HTMLInputElement
    let view: RenderResult<OfferSearchModal, typeof import('@testing-library/dom/types/queries')>

    beforeEach(() => {
        currentModalSettings.title = title
        currentModalSettings.body = body
        currentModalSettings.buttonTextConfirm = undefined
        currentModalSettings.buttonTextCancel = undefined
        currentModalSettings.meta = {
            onClose,
        }
        view = render(OfferSearchModal, {
            props: getComponentsProperties(),
        })
        searchInput = vScreen.getByTestId('search-input') as HTMLInputElement
    })

    describe('Structure', () => {
        it('should have match snapshot', () => {
            expect(view.container).toMatchSnapshot()
        })

        it('should contains a title', () => {
            const titleElement = vScreen.getByText(title)
            expect(titleElement).toBeInTheDocument()
        })

        it('should contains a body', () => {
            const bodyElement = vScreen.getByText(body)
            expect(bodyElement).toBeInTheDocument()
        })

        it('should NOT have a cancel button', () => {
            try {
                vScreen.getByText('cancel-button')
                assert.fail('Cancel button should not be present')
            } catch (e) {
            }
        })

        it('should NOT have a confirm button', () => {
            try {
                vScreen.getByText('continue-button')
                assert.fail('Continue button should not be present')
            } catch (e) {
            }
        })
    })

    describe('Behavior', () => {

        describe('when the search gave some results', () => {

            const offer1 = mock<OfferSummaryPresentation>({
                id: 'offer-1',
            })

            beforeEach(async () => {
                fetchSearchedOffersResult.mockResolvedValue(mock<PaginationPresentation<OfferSummaryPresentation>>({
                    items: [
                        offer1,
                        mock<OfferSummaryPresentation>({
                            id: 'offer-2',
                        }),
                        mock<OfferSummaryPresentation>({
                            id: 'offer-3',
                        }),
                    ],
                }))
                await fireEvent.input(searchInput, {target: {value: 'search with results'}})
            })

            it('should display offers in search result when there are any', () => {

                expect(vScreen.getByTestId('offer-1')).toBeInTheDocument()
                expect(vScreen.getByTestId('offer-2')).toBeInTheDocument()
                expect(vScreen.getByTestId('offer-3')).toBeInTheDocument()
            })

            it('should make the click handler of offer cards call the response method of the modal settings', () => {

                const offerCard = vScreen.getByTestId('offer-1')

                fireEvent.click(offerCard)

                expect(currentModalStoreOnResponse).toHaveBeenCalledWith(offer1)
            })
        })

        describe('otherwise', () => {

            it('should call the close function passed on through the meta of the modal settings when close modal button is clicked', async () => {

                const closeModalButton = vScreen.getByTestId('close-modal-button')

                await fireEvent.click(closeModalButton)

                await waitFor(() => expect(onClose).toHaveBeenCalled())
            })

            it('should show the "no match message" when search did not return any result', async () => {

                await fireEvent.input(searchInput, {target: {value: 'no match'}})
                searchedProductResult.set(mock())

                await waitFor(() => expect(vScreen.getByTestId('no-match-message')).toBeInTheDocument())
            })
        })
    })

    function getComponentsProperties() {
        return {
            parent: mock<SvelteComponent>(),
        }
    }

    afterEach(() => {
        view.unmount()
        vi.clearAllMocks()
        vi.clearAllTimers()
    })
})

vi.mock('@skeletonlabs/skeleton', async () => {
    const actual = await import('@skeletonlabs/skeleton')
    const {modalStore} = await import('$mocks/src/lib/frontend/stores/ModalStore')
    return {
        ...actual,
        getModalStore: (): ModalStore => modalStore,
    }
})

vi.mock('$lib/frontend/endpoints/OfferEndpoints', async () => {
    const {fetchSearchedOffersResult} = await import('$mocks/src/lib/frontend/endpoints/OfferEndpoints')
    return {
        fetchSearchedOffersResult,
    }
})

vi.mock('$lib/frontend/core/RxHelper', async () => {
    return {
        throttle: (func: (...args: any) => any, delay: number) : Promise<(...args: any) => void> => Promise.resolve((...args) => func(args)),
    }
})
