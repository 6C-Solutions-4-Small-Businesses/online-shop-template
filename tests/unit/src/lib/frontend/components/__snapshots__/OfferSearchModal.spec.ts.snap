// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`OfferSearchModal component > Structure > should have match snapshot 1`] = `
<body>
  <div>
    <div
      class="w-full md:w-6/12 lg:w-6/12 xl:w-4/12"
    >
      <div
        class="bg-gray-100 w-full flex flex-col gap-7 items-center rounded-md"
      >
        <header
          class="w-full flex justify-between items-center text-md font-bold uppercase bg-green-50 h-14 text-center p-3 text-black rounded-t-md"
        >
          <p>
            Title
          </p>
           
          <button
            class="w-8 h-8 flex items-center justify-center rounded-full"
            data-testid="close-modal-button"
          >
            <svg
              height="1.2em"
              viewBox="0 0 24 24"
              width="1.2em"
            >
              <path
                d="M12 2c5.53 0 10 4.47 10 10s-4.47 10-10 10S2 17.53 2 12S6.47 2 12 2m3.59 5L12 10.59L8.41 7L7 8.41L10.59 12L7 15.59L8.41 17L12 13.41L15.59 17L17 15.59L13.41 12L17 8.41z"
                fill="currentColor"
              />
            </svg>
            <!--&lt;Close-circle&gt;-->
          </button>
        </header>
         
        <p
          class="w-11/12 text-start text-slate-600"
        >
          Body
        </p>
         
        <div
          class="w-full px-4"
        >
          <div
            class="w-full h-10 flex justify-center"
          >
            <div
              class="relative w-full"
            >
              <input
                class="w-full h-full focus:outline-none text-black text-left pl-2 rounded-none border border-l-gray-500 border-t-gray-500 border-b-gray-500"
                data-testid="search-input"
                placeholder="Chercher un produit"
                type="text"
              />
               
            </div>
             
          </div>
          <!--&lt;SearchInput&gt;-->
           
          <div
            class="py-4"
          >
            <p
              class="text-center text-gray-700"
              data-testid="description-message"
            >
              <span>
                Cherchez l'offre que vous d√©sirez modifier.
              </span>
            </p>
          </div>
        </div>
         
      </div>
      <!--&lt;BaseModal&gt;-->
    </div>
    
    <!--&lt;OfferSearchModal&gt;-->
  </div>
</body>
`;
